File: API/0059_Hammer_Candle/HammerCandleStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies
{
	/// <summary>
	/// Hammer Candle strategy that enters long positions when a hammer candlestick pattern appears.
	/// </summary>
	public class HammerCandleStrategy : Strategy
	{
		private readonly StrategyParam<DataType> _candleType;
		private bool _isPositionOpen;

		/// <summary>
		/// Candle type and timeframe used by the strategy.
		/// </summary>
		public DataType CandleType
		{
			get => _candleType.Value;
			set => _candleType.Value = value;
		}

		/// <summary>
		/// Constructor.
		/// </summary>
		public HammerCandleStrategy()
		{
			_candleType = Param(nameof(CandleType), TimeSpan.FromMinutes(5).TimeFrame())
						 .SetDisplay("Candle Type", "Type of candles to use for pattern detection", "General");
		}

		/// <inheritdoc />
		public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
		{
			return new[] { (Security, CandleType) };
		}

		/// <inheritdoc />
		protected override void OnStarted(DateTimeOffset time)
		{
			base.OnStarted(time);

			_isPositionOpen = false;

			// Subscribe to candles
			var subscription = SubscribeCandles(CandleType);

			subscription
				.Bind(ProcessCandle)
				.Start();

			// Setup stop loss/take profit protection
			StartProtection(
				takeProfit: new Unit(2, UnitTypes.Percent),
				stopLoss: new Unit(1, UnitTypes.Percent)
			);

			// Setup chart visualization if available
			var area = CreateChartArea();
			if (area != null)
			{
				DrawCandles(area, subscription);
				DrawOwnTrades(area);
			}
		}

		private void ProcessCandle(ICandleMessage candle)
		{
			// Skip unfinished candles
			if (candle.State != CandleStates.Finished)
				return;

			// Check if strategy is ready to trade
			if (!IsFormedAndOnlineAndAllowTrading())
				return;

			// Already have a position, wait for exit
			if (_isPositionOpen)
				return;

			// Check for hammer pattern:
			// 1. Lower shadow is at least twice the size of the body
			// 2. Small or no upper shadow
			// 3. Current low is lower than previous low (downtrend)

			var bodySize = Math.Abs(candle.OpenPrice - candle.ClosePrice);
			var lowerShadow = Math.Min(candle.OpenPrice, candle.ClosePrice) - candle.LowPrice;
			var upperShadow = candle.HighPrice - Math.Max(candle.OpenPrice, candle.ClosePrice);

			var isHammer =
				lowerShadow > bodySize * 2m &&
				upperShadow < bodySize * 0.5m &&
				candle.ClosePrice > candle.OpenPrice;  // Bullish candle

			if (isHammer)
			{
				// Enter long position on hammer pattern
				BuyMarket(Volume);
				_isPositionOpen = true;

				LogInfo($"Hammer pattern detected. Low: {candle.LowPrice}, Body size: {bodySize}, Lower shadow: {lowerShadow}");
			}
		}

		/// <inheritdoc />
		protected override void OnPositionChanged(PositionChangeType changeType, decimal value)
		{
			base.OnPositionChanged(changeType, value);

			// Reset position flag when position is closed
			if (Position == 0)
				_isPositionOpen = false;
		}
	}
}

File: API/0059_Hammer_Candle/PY/hammer_candle_strategy.py
import clr

clr.AddReference("System.Drawing")
clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan, Math
from System.Drawing import Color
from StockSharp.Messages import UnitTypes, Unit, DataType, ICandleMessage, CandleStates, Sides
from StockSharp.Algo.Strategies import Strategy
from datatype_extensions import *

class hammer_candle_strategy(Strategy):
    """
    Hammer Candle strategy that enters long positions when a hammer candlestick pattern appears.
    
    """
    def __init__(self):
        super(hammer_candle_strategy, self).__init__()
        
        # Initialize internal state
        self._isPositionOpen = False

        # Initialize strategy parameters
        self._candleType = self.Param("CandleType", tf(5)) \
            .SetDisplay("Candle Type", "Type of candles to use for pattern detection", "General")

    @property
    def CandleType(self):
        return self._candleType.Value

    @CandleType.setter
    def CandleType(self, value):
        self._candleType.Value = value

    def OnReseted(self):
        """
        Resets internal state when strategy is reset.
        """
        super(hammer_candle_strategy, self).OnReseted()
        self._isPositionOpen = False

    def OnStarted(self, time):
        """
        Called when the strategy starts. Sets up indicators, subscriptions, and charting.
        
        :param time: The time when the strategy started.
        """
        super(hammer_candle_strategy, self).OnStarted(time)

        # Subscribe to candles
        subscription = self.SubscribeCandles(self.CandleType)
        subscription.Bind(self.ProcessCandle).Start()

        # Setup stop loss/take profit protection
        self.StartProtection(
            takeProfit=Unit(2, UnitTypes.Percent),
            stopLoss=Unit(1, UnitTypes.Percent)
        )
        # Setup chart visualization if available
        area = self.CreateChartArea()
        if area is not None:
            self.DrawCandles(area, subscription)
            self.DrawOwnTrades(area)

        # Position events will be handled in OnPositionReceived

    def ProcessCandle(self, candle):
        """
        Process candle and execute trading logic
        
        :param candle: The candle message.
        """
        # Skip unfinished candles
        if candle.State != CandleStates.Finished:
            return

        # Check if strategy is ready to trade
        if not self.IsFormedAndOnlineAndAllowTrading():
            return

        # Already have a position, wait for exit
        if self._isPositionOpen:
            return

        # Check for hammer pattern:
        # 1. Lower shadow is at least twice the size of the body
        # 2. Small or no upper shadow
        # 3. Current low is lower than previous low (downtrend)

        bodySize = float(Math.Abs(candle.OpenPrice - candle.ClosePrice))
        lowerShadow = float(Math.Min(candle.OpenPrice, candle.ClosePrice) - candle.LowPrice)
        upperShadow = float(candle.HighPrice - Math.Max(candle.OpenPrice, candle.ClosePrice))

        isHammer = (lowerShadow > bodySize * 2 and
                   upperShadow < bodySize * 0.5 and
                   candle.ClosePrice > candle.OpenPrice)  # Bullish candle

        if isHammer:
            # Enter long position on hammer pattern
            self.BuyMarket(self.Volume)
            self._isPositionOpen = True

            self.LogInfo("Hammer pattern detected. Low: {0}, Body size: {1}, Lower shadow: {2}".format(
                candle.LowPrice, bodySize, lowerShadow))

    def OnPositionReceived(self, position):
        """Handle position information when strategy is started."""
        super(hammer_candle_strategy, self).OnPositionReceived(position)
        if self.Position == 0:
            self._isPositionOpen = False

    def CreateClone(self):
        """
        !! REQUIRED!! Creates a new instance of the strategy.
        """
        return hammer_candle_strategy()